File: backend/internal/model/memory.go
package model

import (
	"crypto/sha256"
	"encoding/hex"
	"strings"
	"time"
)

type Category string

const (
	CategoryPreference Category = "preference"
	CategoryIdentity   Category = "identity"
	CategoryGoal       Category = "goal"
	CategoryContext    Category = "context"
)

func (c Category) Valid() bool {
	switch c {
	case CategoryPreference, CategoryIdentity, CategoryGoal, CategoryContext:
		return true
	default:
		return false
	}
}

type Source string

const (
	SourceChat   Source = "chat"
	SourceManual Source = "manual"
	SourceSystem Source = "system"
)

func (s Source) Valid() bool {
	switch s {
	case SourceChat, SourceManual, SourceSystem:
		return true
	default:
		return false
	}
}

type Memory struct {
	ID          int64     `json:"id"`
	UserID      string    `json:"user_id"`
	Content     string    `json:"content"`
	Category    Category  `json:"category"`
	Source      Source    `json:"source"`
	Importance  int       `json:"importance"`
	ContentHash string    `json:"content_hash"`
	IsDeleted   bool      `json:"is_deleted"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

func (m Memory) Clone() Memory {
	return m
}

type CreateMemoryRequest struct {
	UserID     string   `json:"user_id"`
	Content    string   `json:"content"`
	Category   Category `json:"category"`
	Source     Source   `json:"source"`
	Importance int      `json:"importance"`
}

type UpdateMemoryRequest struct {
	UserID     string   `json:"user_id"`
	Content    string   `json:"content"`
	Category   Category `json:"category"`
	Importance int      `json:"importance"`
}

type ListMemoriesQuery struct {
	UserID   string
	Category Category
	SortBy   string
	Order    string
	Page     int
	PageSize int
}

type ListMemoriesResult struct {
	Items    []Memory `json:"items"`
	Total    int      `json:"total"`
	Page     int      `json:"page"`
	PageSize int      `json:"page_size"`
}

type SearchMemoryResult struct {
	Memory         Memory   `json:"memory"`
	RelevanceScore float64  `json:"relevance_score"`
	RecencyScore   float64  `json:"recency_score"`
	FinalScore     float64  `json:"final_score"`
	MatchedTerms   []string `json:"matched_terms"`
}

type SummaryResult struct {
	Preferences []Memory `json:"preferences"`
	Goals       []Memory `json:"goals"`
	Background  []Memory `json:"background"`
	Recent      []Memory `json:"recent"`
}

func NormalizeContent(content string) string {
	return strings.Join(strings.Fields(strings.TrimSpace(content)), " ")
}

func HashContent(content string) string {
	normalized := strings.ToLower(NormalizeContent(content))
	sum := sha256.Sum256([]byte(normalized))
	return hex.EncodeToString(sum[:])
}

File: frontend/src/types/memory.ts
export type Category = "preference" | "identity" | "goal" | "context";
export type Source = "chat" | "manual" | "system";

export interface Memory {
  id: number;
  user_id: string;
  content: string;
  category: Category;
  source: Source;
  importance: number;
  content_hash: string;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
}

export interface ListMemoriesResult {
  items: Memory[];
  total: number;
  page: number;
  page_size: number;
}

export interface ListMemoriesParams {
  user_id: string;
  category?: string;
  sort_by?: "created_at" | "importance";
  order?: "asc" | "desc";
  page?: number;
  page_size?: number;
}

export interface CreateMemoryPayload {
  user_id: string;
  content: string;
  category: Category;
  source: Source;
  importance: number;
}

export interface UpdateMemoryPayload {
  user_id: string;
  content: string;
  category: Category;
  importance: number;
}

export interface DeleteMemoryResult {
  id: number;
}

export interface SearchItem {
  memory: Memory;
  relevance_score: number;
  recency_score: number;
  final_score: number;
  matched_terms: string[];
}

export interface SearchMemoriesParams {
  user_id: string;
  query: string;
  limit?: number;
}

export interface SearchMemoriesResult {
  items: SearchItem[];
  count: number;
  cached: boolean;
}

export interface SummaryResult {
  preferences: Memory[];
  goals: Memory[];
  background: Memory[];
  recent: Memory[];
}

export interface SummaryPayload {
  summary: SummaryResult;
  cached: boolean;
}

export const CATEGORY_OPTIONS: Array<{ label: string; value: Category }> = [
  { label: "Preference", value: "preference" },
  { label: "Identity", value: "identity" },
  { label: "Goal", value: "goal" },
  { label: "Context", value: "context" }
];

export const SOURCE_OPTIONS: Array<{ label: string; value: Source }> = [
  { label: "Chat", value: "chat" },
  { label: "Manual", value: "manual" },
  { label: "System", value: "system" }
];

export const categoryLabelMap: Record<Category, string> = {
  preference: "Preference",
  identity: "Identity",
  goal: "Goal",
  context: "Context"
};

export const sourceLabelMap: Record<Source, string> = {
  chat: "Chat",
  manual: "Manual",
  system: "System"
};

export function formatDateTime(value: string) {
  if (!value) {
    return "-";
  }
  const date = new Date(value);
  if (Number.isNaN(date.getTime())) {
    return value;
  }
  return new Intl.DateTimeFormat("en-US", {
    year: "numeric",
    month: "short",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit"
  }).format(date);
}
